### PR TITLE
Generalize reduce-window padding to support (lo, hi) pairs.

### PR DESCRIPTION
**Pull request by hawkinsp**
This matches the underlying XLA API, and it turns out to simplify the code slightly, too.

### CODE REVIEW

File: jax/lax/lax.py [hunk 1]
```diff
@@ -1115,11 +1115,14 @@ def _reduce_and(operand: Array, axes: Sequence[int]) -> Array:
 
 def reduce_window(operand: Array, init_value: Array, computation: Callable,
                   window_dimensions: Shape, window_strides: Sequence[int],
-                  padding: str) -> Array:
+                  padding: Union[str, Sequence[Tuple[int, int]]]) -> Array:
   """Wraps XLA's `ReduceWindow
   <https://www.tensorflow.org/xla/operation_semantics#reducewindow>`_
```
Comment (gnecula) [inline jax/lax/lax.py:1120]: This docstring is not quite accurate anymore, it seems. XLA reducewindow is documented to take padding as a string, without the option to pass (lo, hi) pairs.

Comment (gnecula) [reply jax/lax/lax.py:1120]: Is it actually important to expose the (lo, hi) API? In this case, `reduce_window` is more like the XLA `ReduceWindowWithGeneralPadding`.

File: jax/lax/lax.py [hunk 2]
```diff
@@ -4878,6 +4862,7 @@ def _select_and_gather_add_batching_rule(
   x = batching.bdim_at_front(x, x_bdim, size)
   window_dimensions = (1,) + window_dimensions
   window_strides = (1,) + window_strides
+  padding = ((0, 0),) + padding
```
Comment (gnecula) [inline jax/lax/lax.py:4865]: I am wondering how did this work before. Was it working only for the VALID padtype?

**Review by gnecula — APPROVED** [review]
LGTM

File: jax/lax/lax.py (hunk 1 above)
Comment (hawkinsp) [reply jax/lax/lax.py:1120]: Yes, in general we always want to expose the full generality of XLA.

We already started using this functionality as a simplification in this PR in the implementation of the cumulative reductions on TPU.

There doesn't appear to be any documentation on ReduceWindowWithGeneralPadding in XLA; that seems to be a flaw in the XLA documentation.

In general I wouldn't call this operation `ReduceWindowWithGeneralPadding`. It's just `ReduceWindow` with fewer special cases and an API that more closely matches convolution. I don't want to mirror something that simply hasn't been cleaned up in XLA's API.

File: jax/lax/lax.py (hunk 2 above)
Comment (hawkinsp) [reply jax/lax/lax.py:4865]: Both `SAME` and `VALID` padding are identical if the window is of size `1`.

File: jax/lax/lax.py (hunk 1 above)
Comment (hawkinsp) [reply jax/lax/lax.py:1120]: Changed the link text to say `ReduceWindowWithGeneralPadding` even though the link still points to the only documentation there is.

**Pull request merged**